src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1368,7 +1377,60 @@ public ValueTask<FlushResult> WritePipeAsync(ReadOnlyMemory<byte> data, Cancella
             else
             {
                 HandleNonBodyResponseWrite();
-                return !firstWrite ? default : Output.FlushAsync(cancellationToken);
+                return default;
+            }
+        }
+
+        private ValueTask<FlushResult> FirstWriteAsync(ReadOnlyMemory<byte> data, CancellationToken cancellationToken)
+        {
+            Debug.Assert(!HasResponseStarted);
+
+            var startingTask = FireOnStarting();
+
+            if (!ReferenceEquals(startingTask, Task.CompletedTask))
+            {
+                return FirstWriteAsyncAwaited(startingTask, data, cancellationToken);
+            }
+
+            var responseHeaders = InitializeResponseFirstWrite(data.Length);
```

Comment:
```csharp
var responseHeaders = InitializeResponseFirstWrite(data.Length);
```
Could also move to `FirstWriteAsyncInternal`?

Reply:
This also makes sense to me.